### PR TITLE
Rule of Cool explosions v0.2: Impressive Pyrotechnics

### PR DESCRIPTION
**Pull request by DeltaEpsilon7787**
#### Summary

SUMMARY: [Feature] "Explosions are made more spectacular; also make it possible to climb solid rock (thus craters too)"

#### Purpose of change
https://tvtropes.org/pmwiki/pmwiki.php/Main/ImpressivePyrotechnics

This is a further improvement to https://github.com/cataclysmbnteam/Cataclysm-BN/pull/1348, with the intention to both improve the appearance of explosions, make them have a bit more instrumental value and address a few issues that were reported.

#### Describe the solution
1. The whole explosion function is rewritten to now be a process with an event queue. Thanks to this, the explosion can animate itself. This addressed the growing complexity of "do_blast_new".
2. Shrapnel functionality is merged with the explosion process as just one of the events. This addresses the duplication of functionality that was necessary to animate the shrapnel.
3. Implement own ways to move entities around within the explosion process. This both removes the dependency on `fling_creature` method that has arcane things in it for mobs and implement a novel feature of items being thrown around as was suggested in https://github.com/cataclysmbnteam/Cataclysm-BN/pull/1348.
4. The explosion is now animated using fire and smoke fields and does in fact leave some smoke afterwards, but not too much. This addresses a reliance on arcane, almost magical `draw_custom_explosion` method which prevented animations from occurring at different speeds.
5. Constants rewritten a bit to be a bit more understandable in how they affect stuff. This addressed a few constants being hard to understand previously.
6. Terrain bashing functionality rewritten again to be a bit more straightforward.
7. Fire is now drawn OVER smoke. Was needed to make the explosion visible, but also incidentally makes an objectively more dangerous field visible to the player outside of explosions.
8. Solid rock and rock walls are now climbable. Addresses the issue of craters not being easily escapable.

#### Describe alternatives you've considered

N/A

#### Testing

Unaddressed issues/complaints.
1. Flung items deal no damage and flung creatures are dealt no damage. This is a a balancing nightmare, but a number of players have reported disappointment over the lack of damage from this and, honestly, it makes sense, for items to be thrown yet do nothing sorta removes a significant player-controlled way to inflict extra damage and set up impromptu shotguns with nails or something.
2. The animation method I use is done entirely within the process and I give no guarantees it doesn't fuck with some other internals. From my testing, only pixel minimap had a critical issue with texture pool so it's temporarily disabled.
3. I suck at C++ and I give no guarantees I didn't fuck up pointers and shit somewhere. Near a review from someone who knows what they're doing.
4. Shrapnel and blasts are hilariously more dangerous to the player than they are to mobs due to how the damage is inflicted, refer to BN's #development channel, search for "incomprehensibly more dangerous". This needs addressing in a separate PR.
5. Due to leaving traces of old system in, some duplicate code still remains.
6. Header files probably need updating.
7. I give no guarantees the old system and the new system produce equivalent damage.
8. Mininuke is still retarded due to using a separate system to place fd_nuke_gas. Improvement would probably be needed to ExplosionProcess to be able to place arbitrary gas.
9. Z-level vision does not get drawn consistently. Currently interrogating Jove about it.

#### Additional context
Barrel bomb (showcasing issue 9 and also that explosions do in fact get drawn on other Z-levels)
https://cdn.discordapp.com/attachments/830916451517857894/1097472879202947072/2023-04-17_15-44-23.mp4

Mininuke (showcasing issue 8)
https://cdn.discordapp.com/attachments/830916451517857894/1097472895485218886/2023-04-17_15-42-15.mp4

### CODE REVIEW

**Comment by DeltaEpsilon7787** [comment]
Explosion-related tests fail. Ass to fix.

**Review by scarf005 — COMMENTED** [review]
looks sick. some minor nitpick, could you use [constexpr][constexpr], [unpacking][unpack] and [list initialization][list-init] for better readability?

example usage is at [godbolt][godbolt].

[constexpr]: https://en.cppreference.com/w/cpp/language/constexpr
[list-init]: https://en.cppreference.com/w/cpp/language/list_initialization
[unpack]: https://en.cppreference.com/w/cpp/language/structured_binding
[godbolt]: https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:56,endLineNumber:7,positionColumn:56,positionLineNumber:7,selectionStartColumn:56,selectionStartLineNumber:7,startColumn:56,startLineNumber:7),source:'%23include+%3Cstring%3E%0A%23include+%3Cvector%3E%0A%23include+%3Cfmt/core.h%3E%0A%23include+%3Cfmt/format.h%3E%0Ausing+namespace+fmt::literals%3B%0A%0A//+https://en.cppreference.com/w/cpp/language/constexpr%0Aconstexpr+int+foo+%3D+1234%3B%0A%0Aauto+return_pair()+-%3E+std::pair%3Cint,+std::string%3E+%7B%0A++++return+%7B+foo,+%22hello%22+%7D%3B+//+returns+can+be+brace-initialized:+https://en.cppreference.com/w/cpp/language/return%0A%7D%0A%0Aauto+main()+-%3E+int+%7B%0A++++const+auto%26+%5B+amount,+name+%5D+%3D+return_pair()%3B%0A++++%0A++++fmt::println(%22amount+%3D+%7Bamount%7D,+name+%3D+%7Bname%7D%22,+%22amount%22_a+%3D+amount,+%22name%22_a+%3D+name)%3B%0A%0A++++//+https://en.cppreference.com/w/cpp/language/list_initialization%0A++++const+auto+mult_table+%3D+std::vector%3Cstd::pair%3Cint,+int%3E%3E%7B%0A++++++++%7B+1,+2+%7D,+%7B+2,+4+%7D,+%7B+3,+6+%7D,+%7B+4,+8+%7D,+%7B+5,+10+%7D,+%7B+6,+12+%7D%0A++++%7D%3B%0A++++//+https://en.cppreference.com/w/cpp/language/structured_binding%0A++++for+(+const+auto+%5B+a,+b+%5D+:+mult_table+)+%7B%0A++++++++fmt::println(%22%7Ba%7D+*+2+%3D+%7Bb%7D%22,+%22a%22_a+%3D+a,+%22b%22_a+%3D+b)%3B%0A++++%7D%0A%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:49.59578377332948,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:g122,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:fmt,ver:trunk)),options:'-O3+-std%3Dc%2B%2B20',source:1,stdinPanelShown:'1',tree:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+gcc+12.2+(C%2B%2B,+Editor+%231)',t:'0')),k:50.404216226670535,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:clang1500,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,libs:!((name:fmt,ver:trunk)),options:'-O3+-std%3Dc%2B%2B20',source:1,stdinPanelShown:'1',tree:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+clang+15.0.0+(C%2B%2B,+Editor+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50.404216226670535,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

**Comment by DeltaEpsilon7787** [comment]
> looks sick. some minor nitpick, could you use [constexpr](https://en.cppreference.com/w/cpp/language/constexpr), [unpacking](https://en.cppreference.com/w/cpp/language/structured_binding) and [list initialization](https://en.cppreference.com/w/cpp/language/list_initialization) for better readability?
> 
> example usage is at [godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:56,endLineNumber:7,positionColumn:56,positionLineNumber:7,selectionStartColumn:56,selectionStartLineNumber:7,startColumn:56,startLineNumber:7),source:'%23include+%3Cstring%3E%0A%23include+%3Cvector%3E%0A%23include+%3Cfmt/core.h%3E%0A%23include+%3Cfmt/format.h%3E%0Ausing+namespace+fmt::literals%3B%0A%0A//+https://en.cppreference.com/w/cpp/language/constexpr%0Aconstexpr+int+foo+%3D+1234%3B%0A%0Aauto+return_pair()+-%3E+std::pair%3Cint,+std::string%3E+%7B%0A++++return+%7B+foo,+%22hello%22+%7D%3B+//+returns+can+be+brace-initialized:+https://en.cppreference.com/w/cpp/language/return%0A%7D%0A%0Aauto+main()+-%3E+int+%7B%0A++++const+auto%26+%5B+amount,+name+%5D+%3D+return_pair()%3B%0A++++%0A++++fmt::println(%22amount+%3D+%7Bamount%7D,+name+%3D+%7Bname%7D%22,+%22amount%22_a+%3D+amount,+%22name%22_a+%3D+name)%3B%0A%0A++++//+https://en.cppreference.com/w/cpp/language/list_initialization%0A++++const+auto+mult_table+%3D+std::vector%3Cstd::pair%3Cint,+int%3E%3E%7B%0A++++++++%7B+1,+2+%7D,+%7B+2,+4+%7D,+%7B+3,+6+%7D,+%7B+4,+8+%7D,+%7B+5,+10+%7D,+%7B+6,+12+%7D%0A++++%7D%3B%0A++++//+https://en.cppreference.com/w/cpp/language/structured_binding%0A++++for+(+const+auto+%5B+a,+b+%5D+:+mult_table+)+%7B%0A++++++++fmt::println(%22%7Ba%7D+*+2+%3D+%7Bb%7D%22,+%22a%22_a+%3D+a,+%22b%22_a+%3D+b)%3B%0A++++%7D%0A%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:49.59578377332948,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:g122,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:fmt,ver:trunk)),options:'-O3+-std%3Dc%2B%2B20',source:1,stdinPanelShown:'1',tree:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+gcc+12.2+(C%2B%2B,+Editor+%231)',t:'0')),k:50.404216226670535,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:clang1500,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,libs:!((name:fmt,ver:trunk)),options:'-O3+-std%3Dc%2B%2B20',source:1,stdinPanelShown:'1',tree:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+clang+15.0.0+(C%2B%2B,+Editor+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50.404216226670535,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

Changes implemented.

File: src/explosion.cpp [hunk 1]
```diff
@@ -154,461 +158,867 @@ explosion_data load_explosion_data( const JsonObject &jo )
...
-                blasted[critter] += res_dmg;
+void ExplosionProcess::project_shrapnel( const tripoint position )
+{
+    map &here = get_map();
+
+    if( is_occluded( center, position ) ) {
+        return;
+    }
+
+    projectile fragment = shrapnel.value();
```
Comment (scarf005) [inline src/explosion.cpp:552]: it'd be nice to add a comment that it's implied that it's guaranteed to have value

**Review by scarf005 — CHANGES_REQUESTED** [review]
some more nitpicks discussed in discord. not sure what's causing the segfault.

**Review by scarf005 — APPROVED** [review]
https://user-images.githubusercontent.com/54838975/233840388-ae538329-ff44-4174-aeca-eb6bc2874fc4.mp4
(video is speed up by 4x)

looks good to me.
some of furthur nitpicks that could be addressed in v0.3:

flamegraph

- animation is about 10x slower than previous iteration. maybe it's due to large amount of redraws.
- items flinging feels a little bit messy. maybe increasing their chance of destruction or reducing thier fling range could help it.
- it'd be nice to be able to provide an option to use old explosion sprites instead of fire sprites for explosion.

v0.1
https://user-images.githubusercontent.com/54838975/233841521-3d3cd6cb-bed1-4548-9377-b49eb6ec4533.mp4
v0.2
https://user-images.githubusercontent.com/54838975/233841519-f3ba3793-70c0-46d9-a33e-80c53268b8a0.mp4

**Pull request merged**